larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,483 @@
+            continue;
+        }
+        
+        const ClusterList reclusterClusterList(1, clusterList3D.front());
+        std::string clusterListToSaveName, clusterListToDeleteName;
+
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
+        PandoraContentApi::InitializeFragmentation(*this, reclusterClusterList, clusterListToDeleteName, clusterListToSaveName));
+
+        ClusterList newClustersList;
+        for(auto list: minimumFigureOfMeritCaloHitListsVector)
+        {
+            const Cluster *pCluster = nullptr;
+            PandoraContentApi::Cluster::Parameters parameters;
+            parameters.m_caloHitList = *list;
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Cluster::Create(*this, parameters, pCluster));
+            newClustersList.push_back(pCluster);
+        }
+
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, clusterListToSaveName, clusterListToDeleteName));
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->RebuildPfo(pShowerPfo, newClustersList));
+        newCaloHitListsVector.clear();
+    }
+    //If there are any pfos that did not change after reclustering procedure, move them from list called ShowerParticles3D into list of new pfos after reclustering
+    if(unchangedPfoList.size()>0)
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<PfoList>(*this, m_pfoListName, m_PfosForReclusteringListName,  unchangedPfoList));
+
+    //Save list of Pfos after reclustering into list called ShowerParticles3D
+    const PfoList *pNewPfoListAllAfterReclustering;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList<PfoList>(*this, m_PfosForReclusteringListName, pNewPfoListAllAfterReclustering));
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Pfo>(*this, m_PfosForReclusteringListName, m_pfoListName));
+
+    //Set current list to be the same as before reclustering
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Pfo>(*this, initialPfoListName));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::BuildNewTwoDClusters(const Pfo *pPfoToRebuild, ClusterList &newClustersList)
+{
+    ClusterList clusterList2D;
+    LArPfoHelper::GetTwoDClusterList(pPfoToRebuild, clusterList2D);
+
+    for(const Cluster *const pTwoDCluster : clusterList2D)
+    {
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RemoveFromPfo(*this, pPfoToRebuild, pTwoDCluster));
+        HitType hitType = LArClusterHelper::GetClusterHitType(pTwoDCluster);
+        std::string clusterListName(hitType == TPC_VIEW_U ? "ClustersU" : hitType == TPC_VIEW_V ? "ClustersV" : "ClustersW");
```

Comment:
I think these should be configurables, it's a bit inconvenient, but given these list names can be changed in XML I think it's more robust to allow for that.